src/NSwag.CodeGeneration.CSharp/Templates/Client.Class.liquid

```diff
@@ -58,6 +58,53 @@
     partial void PrepareRequest({{ HttpClientType }} client, System.Net.Http.HttpRequestMessage request, System.Text.StringBuilder urlBuilder);
     partial void ProcessResponse({{ HttpClientType }} client, System.Net.Http.HttpResponseMessage response);
 
+    protected struct ObjectResponseResult<T>
+    {
+        public ObjectResponseResult(T responseObject, string responseText)
+        {
+            this.ResponseObject = responseObject;
+            this.ResponseText = responseText;
+        }
+
+        public T      ResponseObject { get; }
+        public string ResponseText   { get; }
+    }
+
+    protected virtual async System.Threading.Tasks.Task<ObjectResponseResult<T>> ReadObjectResponseAsync<T>(System.Net.Http.HttpResponseMessage response, System.Collections.Generic.IReadOnlyDictionary<string,System.Collections.Generic.IEnumerable<string>> responseHeadersAsDictionary)
+    {
+#if DEBUG
+        string responseText = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+        try
+        {
+            T typedBody = Newtonsoft.Json.JsonConvert.DeserializeObject<T>(responseText, this.JsonSerializerSettings);
+            return new ObjectResponseResult<T>(typedBody, responseText);
+        }
+        catch (Newtonsoft.Json.JsonException jsonEx)
+        {
+            string message = "Could not deserialize the response body string as " + typeof(T).FullName + ".";
+            throw new {{ ExceptionClass }}(message: message, (int)response.StatusCode, objectBodyAsString, responseHeadersAsDictionary, jsonEx);
+        }
+#else
+        try
+        {
+            using (System.IO.Stream responseStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false))
+            using (System.IO.StreamReader streamReader = new System.IO.StreamReader(responseStream))
+            using (Newtonsoft.Json.JsonTextReader jsonTextReader = new Newtonsoft.Json.JsonTextReader(streamReader))
+            {
+                Newtonsoft.Json.JsonSerializer serializer = Newtonsoft.Json.JsonSerializer.Create(this.JsonSerializerSettings);
```

Comment:
Maybe we should not recreate the serializer for each request?

Reply:
This is a bit tricky. While `JsonSerializer` is thread-safe, `JsonTextReader` is not, and the `JsonSerializerSettings` object is also directly exposed via the `JsonSerializerSettings { get { ... } }` property. See here: https://stackoverflow.com/questions/36186276/is-the-json-net-jsonserializer-threadsafe - I note this issue exists in the existing NSwag-generated code too.

I did try to use a single instance of `JsonSerializer`, however the `JsonSerializerSettings` property makes use of a `Lazy<T>` field to store it, and creating two objects from a single `Lazy<T>` is non-trivial.

I don't think there are any good solutions while we use `Lazy<T>` and have an exposed property at the same time.

Reply:
I personally would define set up the implicit rule that the settings cannot be changed when the first operation method is called only before when setting up the client... create a new client when you need different settings. With this rule we could reuse the serializer... The serialization settings are only exposed to use in external serializations not to modify it - but there is not a read-only version of the settings that's why the mutable settings have to be exposed... (public settings property option)